src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp

```diff
@@ -189,44 +189,106 @@ namespace OpenLoco::World::MapGenerator
-        regs.ebx = surfaceStyle;
-        call(0x0046A379, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles near water
+        auto seaLevel = getGameState().seaLevel;
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height > seaLevel)
+                continue;
+
+            for (auto lookaheadPos : getClampedRange(pos, pos + TilePos2(50, 50)))
+                heightMap.setMarker({ lookaheadPos.x, lookaheadPos.y });
+        }
+
+        // Apply surface style to tiles that have *not* been marked
+        applySurfaceStyleToMarkedTiles(heightMap, surfaceStyle, false);
     }
 
     // 0x0046A439
     static void generateTerrainNearWater(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A439, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles near water
+        auto seaLevel = getGameState().seaLevel;
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height < seaLevel)
+                continue;
+
+            for (auto lookaheadPos : getClampedRange(pos, pos + TilePos2(50, 50)))
+                heightMap.setMarker({ lookaheadPos.x, lookaheadPos.y });
+        }
+
+        // Apply surface style to tiles that have been marked
+        applySurfaceStyleToMarkedTiles(heightMap, surfaceStyle, true);
     }
 
     // 0x0046A5B3
     static void generateTerrainOnMountains(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A5B3, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles above mountain level
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height < 27)
+                continue;
+
+            for (auto lookaheadPos : getClampedRange(pos, pos + TilePos2(24, 24)))
+                heightMap.setMarker({ lookaheadPos.x, lookaheadPos.y });
+        }
+
+        // Apply surface style to tiles that have been marked
+        applySurfaceStyleToMarkedTiles(heightMap, surfaceStyle, true);
     }
 
     // 0x0046A4F9
     static void generateTerrainFarFromMountains(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A4F9, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles above mountain level
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height < 26)
```

Comment:
should this be 26 or 27 (or should the above one be 26?

Reply:
It surprised me as well, but that's what vanilla uses in both cases yes.